Python/bytecodes.c

```diff
@@ -2804,7 +2804,10 @@ dummy_func(
         replaced op(_FOR_ITER, (iter -- iter, next)) {
             /* before: [iter]; after: [iter, iter()] *or* [] (and jump over END_FOR.) */
             PyObject *iter_o = PyStackRef_AsPyObjectBorrow(iter);
-            PyObject *next_o = (*Py_TYPE(iter_o)->tp_iternext)(iter_o);
+            PyObject *next_o = NULL;
+            if (PyIter_Check(iter_o)) {
```

Comment:
Calling `PyIter_Check` feels a little wasteful. Instead, I'd assign `(*Py_TYPE(iter_o)->tp_iternext)` to a variable and check if it's NULL. If it is NULL, we should raise a TypeError. Your code instead makes it immediately end the loop, which doesn't feel like the right behavior.

Reply:
Hmm, I've tried to add a TypeError 
Such piece of code is provided:

```
            PyTypeObject *type = Py_TYPE(iter_o);
            iternextfunc iternext = type->tp_iternext;
            if (iternext == NULL) {
                _PyErr_Format(tstate, PyExc_TypeError,
                              "'for' requires an object with "
                              "__iter__ method, got %.100s",
                              type->tp_name);
                DECREF_INPUTS();
                ERROR_IF(true, error);
            }
            PyObject *next_o = (*iternext)(iter_o);
```

And there is a temporarily result of my test case:

```
-> % ./python -m unittest -v test.test_generators.GeneratorTest.test_issue125038
test_issue125038 (test.test_generators.GeneratorTest.test_issue125038) ... ERROR

======================================================================
ERROR: test_issue125038 (test.test_generators.GeneratorTest.test_issue125038)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mikhail.efimov/projects/cpython/Lib/test/test_generators.py", line 274, in test_issue125038
    l = list(g)
  File "/home/mikhail.efimov/projects/cpython/Lib/test/test_generators.py", line 272, in <genexpr>
    g = (x for x in range(10))
                    ~~~~~^^^^
TypeError: 'for' requires an object with __iter__ method, got range

----------------------------------------------------------------------
Ran 1 test in 0.001s
```

It seems like such a message doesn't provide any clarity.
Do you have any better suggestions about error message?

Reply:
That seems like a pretty good error message, much better than silently stopping the loop.